RIOT-OS/RIOT
# pkg/opendsme: add initial support for IEEE 802.15.4 DSME time-slotted MAC

### Contribution description
This PR adds support for [openDSME](http://opendsme.org).
DSME is a time-slotted MAC layer (similar to TSCH), which supports deterministic communication.
It offers some promising features such as:
- Topology agnostic (peer-to-peer, star, mesh, etc).
- By design, it's possible to extend the network just adding more coordinators near the edge.
- Automatic resolution of beacon collision
- Transmission of frames using either CSMA-CA or multi-channel GTS (guaranteed time slot)
- Built-in slot negotiation

Similar to TSCH, such a MAC improves PRR drastically (with the cost of some latency) and enables low-power operation.

This work is still WIP, although the basic support is already running. This PR includes:
- Basic openDSME support as a pkg
- Implementation of GNRC interface for openDSME (IP support will be a follow up)
- An example application

~Some commits (labeled as "REMOVE_ME") will be handled in separate PRs~
Tested with `nrf52840`. The cc2538 needs a small change to remove the beacon filter.




### Testing procedure

1. Flash 2 boards (or more) with `examples/opendsme`.
2. Choose one boards as the PAN coordinator and run:
```
ifconfig 3 pan_coord
ifconfig 3 up
ifconfig     <--- Should display "Link: up" and show the short address.
```
3. Join the DSME network in the shell of the other nodes. Wait a fraction of a second and get the status and address.
```
ifconfig 3 up
ifconfig    <--- Should display "Link: up" and show the short address.
```
4. Use the `gts` option to enable GTS transmission (slotted). By default, the option is unset.
- Send data using CSMA-CA
```
ifconfig 3 -gts
txtsnd 3 <l2_addr> login 
```
- Send data using GTS
```
ifconfig 3 gts
txtsnd 3 <l2_addr> login 
```

## Example session

PAN coordinator
```
2022-06-01 15:22:43,584 # main(): This is RIOT! (Version: 2022.07-devel-595-g61467b-pr/opendsme_v3)
ifconfig 3 pan_coord
2022-06-01 15:24:40,160 # ifconfig 3 pan_coord
2022-06-01 15:24:40,162 # success: set option
> ifconfig 3 up
2022-06-01 15:24:48,087 # ifconfig 3 up
2022-06-01 15:24:48,090 # This node is PAN coordinator
> 2022-06-01 15:25:23,347 # PKTDUMP: data received:
2022-06-01 15:25:23,352 # ~~ SNIP  0 - size:  12 byte, type: NETTYPE_NETIF (-1)
2022-06-01 15:25:23,355 # if_pid: 0  rssi: -32768  lqi: 0
2022-06-01 15:25:23,356 # flags: 0x0
2022-06-01 15:25:23,357 # src_l2addr: BD:3F
2022-06-01 15:25:23,359 # dst_l2addr: EB:82
2022-06-01 15:25:23,364 # ~~ SNIP  1 - size:   4 byte, type: NETTYPE_UNDEF (0)
2022-06-01 15:25:23,366 # 00000000  52  49  4F  54
2022-06-01 15:25:23,371 # ~~ SNIP  2 - size:   9 byte, type: NETTYPE_UNDEF (0)
2022-06-01 15:25:23,374 # 00000000  41  A8  39  23  00  82  EB  3F  BD
2022-06-01 15:25:23,378 # ~~ PKT    -  3 snips, total size:  25 byte
```

Child:
```
Welcome to pyterm!
Type '/exit' to exit.
ifconfig 3 up
2022-06-01 15:25:07,341 # ifconfig 3 up
> ifconfig
2022-06-01 15:25:09,228 # ifconfig
2022-06-01 15:25:09,232 # Iface  3  HWaddr: BD:3F  Link: up 
2022-06-01 15:25:09,234 #           L2-PDU:127  
2022-06-01 15:25:09,235 #           
ifconfig 3 gts
2022-06-01 15:25:17,318 # ifconfig 3 gts
2022-06-01 15:25:17,319 # success: set option
txtsnd 3 EB:82 RIOT
2022-06-01 15:25:23,184 # txtsnd 3 EB:82 RIOT
> 

```



### Issues/PRs references
#15313 
Depends on ~#18283, #18284~
